### DISCUSSION TITLE
Compile time (train_step) is too long when using non-`scan` layers.

### DISCUSSION THREAD

**Discussion by carlesoctav** [marin-community/levanter#1279] [Q&A]
I just found (while debugging my own compile times) a fact: **non-`scan` layers cause the compiler to unroll computation across layers**, which increases compile time.
I was building a small library on top of Equinox and noticed the issue. I compared my implementation to Flax’s BERT (flaxformer) on maskedlm task and got the following:

## Compile time

| Variant | With Dropout | Without Dropout |
| ------- | -----------: | --------------: |
| Equinox |        216 s |            67 s |
| Flax    |         78 s |            38 s |

## Runtime (per step)

| Variant | With Dropout | Without Dropout |
| ------- | -----------: | --------------: |
| Equinox |       120 ms |          100 ms |
| Flax    |       150 ms |           90 ms |

with the following config:

```python
config = BertConfig(
    vocab_size=30522,
    hidden_size=768,
    num_hidden_layers=12,
    num_attention_heads=12,
    intermediate_size=3072,
    max_position_embeddings=512,
    type_vocab_size=2,
    hidden_dropout_prob=0.1 or 0,
    attention_probs_dropout_prob=0.1 or 0,
    _attn_implementation="eager",
)
```

Step times are comparable, but compile times aren’t. I then ran Levanter with a GPT-2 of roughly the same parameter count and saw ~30 s compile time. From that I learned Levanter stacks layers with `jax.lax.scan`, and how `scan` interacts with `jax.jit`.

After porting a `scan`-based layer into my project, my compile time matched Levanter’s GPT-2 numbers!

Next, I tested a non-`scan` “BlockSeq” variant. Gemma 3 exposes a flag to switch off `scan` and use block sequencing:

```yaml
data:
  id: dlwh/wikitext_103_detokenized
model:
  type: gemma
  hidden_dim: 768
  seq_len: 512
  intermediate_dim: 3072
  num_heads: 12
  num_layers: 12
  head_dim: 64
  num_kv_heads: 12
  vocab_size: 30522
  scan_layers: False

trainer:
  mp: bfloat16
  num_train_steps: 100
  checkpointer:
    keep:
      - every: 50
    save_interval: 5m
  per_device_parallelism: -1
  train_batch_size: 64
  tensor_parallel_axes: ["mlp", "heads"]
  fsdp_axis: "embed"
  batch_axis: "batch"
```

With `scan_layers: False`, compile time jumped to ~130 s. (I think Gemma disables dropout by default, so I couldn’t do a dropout-on/off comparison here. Even though this is a different model, with similar parameter counts I’d expect compile time trends comparable to Flax/Equinox—correct me if I’m wrong.)

Log excerpt:
```
2025-10-28T16:16:09 - 0 - levanter.data.loader - loader.py:261 - INFO :: Prefetch wasn't fast enough: 0.901.
2025-10-28T16:18:46 - 0 - levanter.trainer - trainer.py:742 - INFO :: DEBUGPRINT[72]: trainer.py:741: diff=133.38703838200308
# diff was compile time
```

Is this just an expected trade-off we accept when using pytree-based module (instead of doing f(params, x)) (and solve with `scan`), What puzzles me is that Flax’s BERT seems to achieve good compile times without `scan` (I used the Flaxformer BERT implementation).

Notes on how I measured compile time:

I added the following around the lowering/compilation step, and ran everything on a single TPU v4 VM. The Flax and Equinox runs were doing DDP during timing; I’m not 100% sure about the Levanter Gemma-3 run in that respect.

```python
def _maybe_save_jaxpr(self, name: str, fn, *args, **kwargs):
    logged = False
    if self.config.log_jaxprs and name not in self._logged_jaxprs:
        jaxpr, _, _ = eqx.filter_make_jaxpr(fn)(*args, **kwargs)
        pretty = jaxpr.pretty_print(name_stack=True, use_color=False)
        self.write_artifact(f"{name}.jaxpr.txt.gz", pretty, type="jaxpr")
        logged = True

    if self.config.log_xla_hlo and name not in self._logged_jaxprs:
        hlo = fn.lower(*args, **kwargs).as_text("stablehlo")
        self.write_artifact(f"{name}.hlo.txt", hlo, type="hlo")
        with open(f"{name}.hlo.txt", "w") as f:
            f.write(hlo)

        start = time.monotonic()
        fn.lower(*args, **kwargs).compile()
        diff = time.monotonic() - start
        logger.info(f"DEBUGPRINT[72]: trainer.py:741: diff={diff}")
        logged = True

    if logged:
        self._logged_jaxprs.add(name)

    return fn(*args, **kwargs)
```

**Comment by dlwh** [comment]
yep this is a thing and the main reason to use scan layers. There's honestly not a very good reason to other than compile times AFAIK. You can use [compilation caching](https://docs.jax.dev/en/latest/persistent_compilation_cache.html) but it imposes some constraints on what can go inside jitted regions.

**Reply by carlesoctav** [reply]
Any thoughts on why this is not that bad in flax (even though it doesnt use scan layers)? is it due to the simpler structure (params just pure dict), but I'M curious if therea are other reasons.

**Comment by carlesoctav** [comment]
Another observation from my tests:

Initializing the model takes about 80 seconds to finish (again, this is with non-scan layers).

I’ve noticed a similar issue in my own library. We initialize the module under `jit` so that parameters can be sharded according to the partition spec (we do this so arrays are initialized directly on TPUs, not just a single device).

I kind of solved this by making RNG/key splitting lazy, and using `fold_in` (though I’m not sure if using `fold_in` instead of `split` actually helps), similar to Flax. However, this required creating another object to track the counter and original seeds.

Later, I decided to use a simple splitting key and avoid initializing the module under `jit`.

If I know I can't load the module on a single device, I create an abstract module and reinitialize each array one by one, using `device_put` or `lax.with_sharding_constraint` after each reinitialization. This should help avoid potential OOM issues when initializing on a single device.

Technically, you don’t need to do this if you’re using scan layers, just an observation!

```
2025-10-29T02:19:46 - 0 - levanter.checkpoint - checkpoint.py:602 - WARNING :: No checkpoints found in checkpoints/o82la4tv
2025-10-29T02:19:48 - 0 - levanter.store.cache - cache.py:303 - INFO :: Loading cache from cache/train
2025-10-29T02:19:49 - 0 - levanter.data.text - text.py:1010 - WARNING :: Defaulting to linear permutation for shuffling. This will change to Feistel in the future.
2025-10-29T02:19:50 - 0 - levanter.store.cache - cache.py:303 - INFO :: Loading cache from cache/validation
DEBUGPRINT[55]: train_lm.py:154: init_model diff=74.37349540297873
```

**Comment by carlesoctav** [comment]
also i can make the compile time jump to 500s even with only with 220M params model.
try to turnoff scan_layers, gradient_checkpointing, use vanilla attention, set attention_dropout, embed_dropout, and resid_dropout, and set the mp to float32 ...

**Comment by carlesoctav** [comment]
After trying to recreate this with Flax (without Flaxformer) and NNX, I can say that this kind of long compile time (around 200s) is quite common between these implementations (equinox, levanter, flax linen, nnx). But, I still don’t know what makes the Flaxformer implementation different, aside from the fused embedding and fused QKV layers. :(

**Reply by dlwh** [reply]
this seems like a thing to ask the jax team at this point. I just use scan layers
